File: Problem 6/sentimental-credit/credit.py
from cs50 import get_int
# TODOchecksum(card) == True and
def main():
    while True:
        card = get_int("Number: ")

        if amex(card) == True and checksum(card) == True:
            print("AMEX\n")
        elif master(card) == True and checksum(card) == True:
            print("MASTERCARD\n")
        elif visa(card) == True and checksum(card) == True:
            print("VISA\n")
        else:
            print("INVALID\n")
        break

def checksum(card):
    sum = 0
    # temp store number that need to *2
    temp = 0
    for i in range(len(str(card))):
            if i % 2 == 0:
                sum += card % 10
            else:
                temp = 2*(card % 10)
                sum += (temp//10) + (temp%10)
            card //= 10

    if (sum % 10) == 0:
        return True

def amex(card):
    if (len(str(card)) == 15) and (int(str(card)[:2]) == 34 or int(str(card)[:2]) == 37):
        return True

def master(card):
    if (len(str(card)) == 16) and (int(str(card)[:2]) >=51 and int(str(card)[:2]) <=55):
        return True

def visa(card):
    if (len(str(card)) == 13 or len(str(card)) == 16) and (int(str(card)[:1]) == 4):
        return True


main()

File: Problem 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>


int main(void)
{
    //get number from user
    long num = get_long("Number: ");

    // get digit length
    int len = floor(log10(num)) + 1;

    // convert to array
    long numbackup1 = num;
    int numbers[len];
    int n = len-1;
    while (numbackup1 > 0)
    {
        int tempnum = numbackup1 % 10;
        numbackup1 /= 10;
        numbers[n] = tempnum;
        n--;
    }
    // checksum
    // declare variable to store value, temp is value that needs to *2
    long numbackup2 = num;
    int sum = 0;
    int pos = 0;
    while (numbackup2 != 0)
    {
        if(pos % 2 != 0)
        {
            int temp = 2 * (numbackup2 %10);
            if (temp > 9)
            {
                // mod(%) for remainder, divide(/) for the value in tenth place
                sum += (temp % 10) + (temp / 10);
            }
            else {
                sum += temp;
            }
        }
        else
        {
            sum += numbackup2 % 10;
        }
        numbackup2 = numbackup2 / 10;
        pos++;
    }

    if (sum % 10 == 0)
    {
        if (len == 15 && numbers[0] == 3 && (numbers[1] == 4 || numbers[1] == 7))
        {
            printf("AMEX\n");
        }
        else if (len == 16 && numbers[0] == 5 && numbers[1] >= 1 && numbers[1] <= 5)
        {
            printf("MASTERCARD\n");
        }
        else if ((len == 13 || len == 16) && numbers[0] == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

}